### DISCUSSION TITLE
Bitwarden design flaw: Server side iterations

### DISCUSSION THREAD

**Discussion by sschueller** [dani-garcia/vaultwarden#3162] [General]
There was just an article posted on HN regarding this server side iteration being too low in bitwarden: https://palant.info/2023/01/23/bitwarden-design-flaw-server-side-iterations/

What are the server side iterations set too in vaultwarden? The client default appears to be 100k which I guess should be higher as well since OWASP just changed the recommendation to 600k

HN: discussion: https://news.ycombinator.com/item?id=34497898

**Comment by NeoLizzard** [comment]
In vaultwarden you can find your current settings (server side) under "/admin" or you can check the config.json for it's current value and change it there (it is actually the same place /admin == config.json). Or you can set it in the .env (https://github.com/dani-garcia/vaultwarden/blob/main/.env.template) However my question would be - how can we tackle the user settings? Is there a way in vaultwarden to manipulate the default user creation skelleton and set the value to let's say 600k (and yes - I know about the compatibility issues with old devices)

**Comment by BlackDex** [comment]
The client-side iterations are defined in the client. So, that would be the web-vault, or any other client which is able to create an account.
At the moment this is still set to `100000` with the web-vault provided with the Vaultwarden images (Same btw for the Bitwarden Self-Hosted)

**Reply by NeoLizzard** [reply]
I know, that the client-side iterations are defined in the client. But when we generate them with vaultwarden, we are in control of the predefined value that is set at some point (e.g. in web-wault, where no external client has initial control over) - therefore we should be able to hook into the account creation process and raise the default value - the only question in place is ... where? :)

**Comment by NeoLizzard** [comment]
EDIT: Removed my dirty hack, since it was not working as expected and caused account logout - don't apply!

**Reply by NeoLizzard** [reply]
Nevermind - doesn't work - can't login afterwards

**Reply by NeoLizzard** [reply]
The password doesn't fit after it anymore

**Reply by BlackDex** [reply]
I'm already working on fixing this for the Vaultwarden side for the hashed password storage to also be increased.
That will not fix the client-side iterations though! That is something you need to change your self via the web-client, account > security > keys.

**Reply by BlackDex** [reply]
See #3163

**Reply by tessus** [reply]
I am a bit fuzzy as to what will happen, if the server has the iterations set to 350000 and the client uses anything but 350000. The client has all the unencrypted data and the server only ever has encrypted data, so why is there a server side parameter for key iterations? Is this just for the web token?

**Reply by tessus** [reply]
Additionally the paper above mentions that changing the server side iterations does not have any effect and only the client side iterations improve the situation.

**Reply by BlackDex** [reply]
@tessus, the clients send a hash of the password. If you can match that hash via brute-force, you will have the password of the user. That is why on the server side we also run that hash through PBKDF using an x amount of iterations, so that it's not easy to brute-force this. Since it is better to have a higher value then 100_000 this PR changes that default value.

It will also update every user during login with the new default value. Or when the server admin increases (or decreases) this value.

The reason people are still allowed to set it back to 100_000 is that some hardware could have a hard time running this amount of iterations and therefore allow it to set at least to 100_000, but not lower.

That changing it on the server-side doesn't have any effect, doesn't mean we shouldn't of course.

**Reply by tessus** [reply]
If I send a password with X (client) hash iterations the hash is going to be different than the hash that is generated with Y (server) hash iterations. But the server never gets the unencrypted master password, so I am currently at a loss.

Ok, I obviously haven't properly looked at the security architecture. I will have to do that. Until then I will shut up. ;-)

**Reply by BlackDex** [reply]
The clients send a hash based upon your master-password and your mail-address etc.. The hash that is generated by this is always exactly the same. That hash is then sent to the server (Doesn't matter if it is Bitwarden or Vaultwarden). This hash then needs to be checked if that is valid. But, if we would store only the hash as-is, that would mean someone could really simple check that hash from the database, which would be very bad.

That is why on the server-side we run that hash we receive through PBKDF with an x amount of iterations. We validate the stored PBKDF with the newly generated PBKDF and if they match, then the user has provided a valid password and they can continue.

Also, just to provide some more info. Even if someone would generate a valid master-password/mail-address hash them selfs and put that into the database manually, they are not able to decrypt the vault. They are allowed to login though, but the vault is unable to be decrypted and will provide them no useful data.

**Reply by tessus** [reply]
Thanks for the explanation. One thing though:

> The hash that is generated by this is always exactly the same.

This is only true if I don't change the iterations on the cilent. So let's say I change the iterations on the client from X to Y, the hash won't be the same. So the stored value on the server won't match when you use the new client hash to run through the server iterations.

This means that as soon as the client or server hash iteration value changes, the stored data on the server must be changed as well.

**Reply by BlackDex** [reply]
But the clients send the hash, and thus we depend on the client to provide this information.
Also, we do not care how many iterations the client does, as long as that hash is correct.

If the user changes there KDF (which is not done automatically as far as i know) we also receive a new master-password hash, which is then stored. So, there is no issue here.

**Reply by tessus** [reply]
> If the user changes there KDF (which is not done automatically as far as i know) we also receive a new master-password hash, which is then stored. So, there is no issue here.

That's the crucial part. Now it makes sense, and underscores what I mentioned in my previous comment. All good now. Cheers

**Reply by Coronon** [reply]
Please correct me if I am wrong, but we should be able to implement changing the client side interactions for both new and already registered users. For new users we should be able to change the default used on signup (at least on the web version, I don’t know if you can sign up through the app?). For already registered users we could run the change as soon as they login, so that the vault is decrypted (at least on the web version, as we are effectively in control of the JS that handles all the inputs anyway).

or have I forgotten about something here?

**Reply by BlackDex** [reply]
The encryption part is something i want to stay away from on our side actually. Modifying the web-vault on that level is asking for issues in my opinion. I rather have that Bitwarden does this on there side automatically, which they can, and they can push that to the server in which case we can update that info.

**Reply by Coronon** [reply]
Thats understandable, let’s hope that Bitwarden acknowledges this problem soon and pushes changes. It was on the front page of HN, so the community is definitely aware - question being if Bitwarden the company cares

**Reply by BlackDex** [reply]
Well, i think they care, since they have changed the defaults currently to 600k, yet unreleased, but it's in there master branch on both, server, mobile and clients repo.

The only thing i do not know, is if they somehow do an auto-update of the default iterations for the user in some way if there current default is the older default. The problem with that is, you need to know if a user has changed it them self or not. It could be that they have a device which doesn't support a higher amount of iterations without a very large slowdown.

But, I don't know. We will have to wait and see. It shouldn't be that hard for them to implement something like this :).

**Reply by Coronon** [reply]
Thank you for the update, exited to see whats coming :)

**Comment by masterwishx** [comment]
sorry im confused a little with password iterations, in bitwarden site i found they have 600000 on server side + 100000 on client side .
But i had only 100000 on server side and changed to 350000 becose of default for vaultwarden now. and having some second of entering to vault on mobile now ....
can find this option on mobile bitwarden also ....

**Reply by BlackDex** [reply]
There are two locations regarding the itterations.
1. Server-side (Vaultwarden/Bitwarden)
This is how many times the password-hash which is sent by all the clients during login is iterated over before it is stored in the database. The same needs to be done to validate the stored value.

The default for Vaultwarden is 600.000. This was 100.000 before, but is upgrade automatically unless you provided a custom setting your self.

2. Client-side (Bitwarden Clients)
This is how many times the password it self is iterated over which then is used to derive a master-key from.

The default is set by Bitwarden. This was 100.000, but currently is 600.000. This setting however is not updated automatically, and you need to change it your self in the settings.

The amount of iterations means a device needs to loop that many times, which needs CPU. If you have a lower-end device, then it will take a bit longer to login.

**Reply by masterwishx** [reply]
Do you mean password iterations in config.json it's server side (350000 default in env.*)
And on vault.myside.com security->keys->change kdf it's client side?

**Reply by BlackDex** [reply]
The server default is 600.000 not 350.000.
But yes, your statement is correct

**Reply by masterwishx** [reply]
Sorry it's what I see in env.template

**Reply by BlackDex** [reply]
Those are not defaults. Some might be, some might not.
A few items are examples, some others are the opposite of the default, like some booleans.

The defaults are in the `config.rs` file.
https://github.com/dani-garcia/vaultwarden/blob/0b28ab3be101b58cf27b43da14bcf893fa37c1c8/src/config.rs#L471

**Reply by masterwishx** [reply]
OK, I see, so should i change to 600000 also...

**Reply by BlackDex** [reply]
That all depends. For the server side if your server is powerful enough so that it doesn't cause too much slowdown.
The same goes for the client-side. Adjust the values in that way that it is still workable for your devices.

But keep in mind, that the whole reasoning behind this is that, the more iterations, the more time it takes to validate and allow you access, the more time it takes for a hacker to bruteforce search for your password if they have been able to get your hashed data.

**Reply by masterwishx** [reply]
OK, Thanks i will test it.
I have Unraid and a lot of dockers including vaultwarden. On i5 4440 4 core 24GB.
Yesterday changed to 350000 from 100000 both and on my old mobile having a couple of seconds to enter :(

**Reply by NeoLizzard** [reply]
In regards of performance you can try to use a dedicated database in your docker compose.
I have a multitude faster performance in my case for search etc. But I haven't checked if it also affect the encryption\decryption speed.
Based however on the fact, that the hashes need to be get from the database, it seam logic to me that in case of faster serving you may get a better login speed aswell and maybe save a couple of seconds there.

For reference - my current docker composed (change passwords and adjust settings as needed) - Keep in mind, that your previous database will not get migrated into the new one - so export your data prior trying this here, so you can import it in the new database.

```
version: "3.7"
services:
 mariadb:
  image: "mariadb"
  container_name: "mariadb"
  restart: always
  env_file:
   - ".env"
  volumes:
   - "/opt/vaultwarden/data/mariadb/:/var/lib/mysql"
   - "/etc/localtime:/etc/localtime:ro"
  environment:
   - "MYSQL_ROOT_PASSWORD=YOURROOTPASSWORDHERE"
   - "MYSQL_PASSWORD=YOURMYSQLPASSWORDHERE"
   - "MYSQL_DATABASE=vaultwarden"
   - "MYSQL_USER=vaultwarden"

 vaultwarden:
  image: "vaultwarden/server:latest"
  container_name: "vaultwarden"
  restart: always
  env_file:
   - ".env"
  volumes:
   - "/opt/vaultwarden/data/:/data/"
  environment:
   - "DATABASE_URL=mysql://vaultwarden:YOURPASSWORDHERE@mariadb/vaultwarden"
   - "RUST_BACKTRACE=1"
   - "EXTENDED_LOGGING=true"
   - "USE_SYSLOG=false"
   - "LOG_LEVEL=error"
   - "ROCKET_CLI_COLORS=false"
   - "LOG_FILE=/data/vaultwarden.log"
   - "WEBSOCKET_ENABLED=true"
```

**Reply by BlackDex** [reply]
Searching in the Bitwarden clients have nothing to do with a database backend. Nothing is done on the server side regarding this.

Only the initial loading of the vault data could be affected.
It also has nothing to do with password verification, since checking the records if the hash matches doesn't take that long.

**Reply by NeoLizzard** [reply]
Thank you for clarification - it was just an idea from my side. As I wrote prior - haven't tested it.

**Reply by masterwishx** [reply]
> Searching in the Bitwarden clients have nothing to do with a database backend. Nothing is done on the server side regarding this.

Thanks for explaining...

**Reply by masterwishx** [reply]
I'm using vaultwarden with SQLite by default but also have dedicated Mariadb, Postgresql.

**Reply by masterwishx** [reply]
> Only the initial loading of the vault data could be affected.

Sorry for my last question here, do you mean if I move to mariadb from sqllite it will increase initial loading of data?
Becouse I see some latency after entering to vault and loading data on my mobile...

**Reply by masterwishx** [reply]
I changed to Mariadb from sqllite, but didn't saw a faster initial load data in mobile, so changed back to sqllite for now.
I have only one user for now but will have 5-7 famaly users maybe in future, I hope it's good for SQLite?
